docs/proposals/20210335-load-balancer-provider.md

```diff
@@ -0,0 +1,999 @@
+    - Set `Status.ControlPlaneReady` and conditions based on state of `cp`
+
+##### Scenario: Cluster API managed endpoint
+
+User wants Cluster API to manage the endpoint for the cluster control plane.
+
+Preconditions:
+- `Spec.ControlPlane.Endpoint` is empty
+- `Spec.ControlPlane.ProviderRef` is defined
+
+Reconciliation:
+- Fetch the referenced provider CR as `cp`
+  - If `cp.Spec.Endpoint` is empty, return early and wait for triggered re-reconciliation
+  - If `cp.Spec.Endpoint` is non-empty and is not equal to `Spec.ControlPlane.Endpoint`, error
+  - If `cp.Spec.Endpoint` is non-empty and is equal to `Spec.ControlPlane.Endpoint`:
+    - Set `Status.ControlPlaneReady` and conditions based on state of `cp`
+
+#### Changes to ControlPlane provider types:
+
+- Add spec.Endpoint, if not defined it will be populated by the load balancer provider (if defined)
+- Add optional LoadBalancerRef (if needed, based on the provider)
+
+#### New Watches:
+
+- LoadBalancer
+
+##### Example: KubeadmControlPlane
+
+```yaml
+apiVersion: controlplane.cluster.x-k8s.io/v1alpha4
+kind: KubeadmControlPlane
+metadata:
+  name: my-cluster-control-plane
+spec:
+  endpoint:
+    host: test.example.com
+    port: 443
+  loadBalancerRef:
+    apiGroup: cluster.x-k8s.io
+    kind: LoadBalancer
+    name: my-cluster-control-plane-lb
+```
+
+#### Changes to ControlPlane provider reconciliation:
+
+##### Example: KubeadmControlPlane
+
+![Kubeadm Control Plane Reconciliation Flow](images/load-balancer/kcp-reconciliation.png)
+
+###### Scenario: Load Balancer specified, externally managed endpoint
+
+User wants to provide an externally managed endpoint for use when creating a cluster, but wants Cluster API to manage the load balancer for the control plane.
+
+Preconditions:
+- `Spec.Endpoint` is non-empty (could be populated by the Cluster controller)
+- `Spec.LoadBalancerRef` is defined
+
+Reconciliation:
+TODO
+
+###### Scenario: Load Balancer specified, provider managed endpoint
+
+User wants Cluster API to manage the endpoint for the control plane when creating a cluster.
+
+Preconditions:
+- `Spec.Endpoint` is empty
+- `Spec.LoadBalancerRef` is defined
+
+Reconciliation:
+TODO
+
+###### Scenario: Load Balancer not specified (fallback behavior)
+
+User wants Cluster API to manage the endpoint for the control plane, but is using a provider that doesn't yet support pluggable load balancers.
+
+Preconditions:
+- `Spec.Endpoint` is empty
+- `Spec.LoadBalancerRef` is nil
+
+Reconciliation:
+TODO
+
+### New Types
+
+#### Load Balancer
+
+A common type to aggregate information between Load Balancer provider implementations and to provide some common reconciliation logic.
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1alpha4
+kind: LoadBalancer
+  metadata:
+    name: my-cluster-control-plane-lb
+spec:
+  clusterName: my-cluster
+  endpoint:
+    host: test.example.com
+    port: 443
+  targetKind:
+    apiGroup: cluster.x-k8s.io
+    kind: Machine
```

Comment:
I don't think we'd want to support multiple `targetKinds`, though this wouldn't exclude someone from combining different machine infrastructure providers, since the target in this case is Machines. This was more to help facilitate implementations that may not be Machine-based (such as a pod-based control plane implementation).